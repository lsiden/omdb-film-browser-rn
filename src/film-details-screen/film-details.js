import React from "react"
import PropTypes from "prop-types"
import { ScrollView } from "react-native"
import { List } from "react-native-elements"

import Poster from "./poster"
import FilmPlot from "./film-plot"
import DetailItem from "./detail-item"
import detailItems from "./detail-items"
import { detailStyles as style } from "styles"

let i = 1
const genKey = () => `item-${i++}`

const FilmDetails = ({ details, onPressPoster }) => (
  <ScrollView style={style.wrapper}>
    <Poster uri={details.Poster} onPress={onPressPoster} />
    <FilmPlot />
    <List>
      {detailItems(details).map(detail => (
        <DetailItem key={genKey()} detail={detail} />
      ))}
    </List>
  </ScrollView>
)

FilmDetails.propTypes = {
  details: PropTypes.object.isRequired,
  onPressPoster: PropTypes.func.isRequired
}

export default FilmDetails
