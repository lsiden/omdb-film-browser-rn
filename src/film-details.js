import React from "react"
import PropTypes from "prop-types"
import { connect } from "react-redux"
import { Text, View, Linking, Image, TouchableOpacity } from "react-native"
import { List, ListItem } from "react-native-elements"

import { viewList } from "./actions"
import { detailStyles } from "./styles"
import { LEFT_TRIANGLE } from "./constants"
import LoadingIndicator from "./components/loading-indicator"

const detailExists = detail => detail && detail !== "N/A"

function OneDetail({ detail }) {
  const { text, cond = true, url = null } = detail

  if (!cond) {
    return null
  }

  const props = {
    style: detailStyles.item,
    title: text,
    hideChevron: true,
  }

  if (detailExists(url)) {
    props.onPress = () => {
      Linking.openURL(url)
    }
    props.hideChevron = false
  }
  return <ListItem {...props} />
}

OneDetail.propTypes = {
  detail: PropTypes.object.isRequired,
}

const imdbUrl = imdbID => `https://www.imdb.com/title/${imdbID}`

const renderBackButton = ({ dispatchViewList }) => (
  <TouchableOpacity onPress={dispatchViewList}>
    <Text style={detailStyles.buttonText}>{LEFT_TRIANGLE}</Text>
  </TouchableOpacity>
)
renderBackButton.propTypes = {
  dispatchViewList: PropTypes.func.isRequired,
}

const DetailHeader = props => {
  const { filmSummary } = props
  return (
    <View style={detailStyles.header}>
      {renderBackButton(props)}
      <View style={detailStyles.titleWrapper}>
        <Text style={detailStyles.title}>{filmSummary.Title}</Text>
        <Text style={detailStyles.year}>{`  (${filmSummary.Year})`}</Text>
      </View>
    </View>
  )
}
DetailHeader.propTypes = {
  filmSummary: PropTypes.object.isRequired,
}

const Poster = ({ filmSummary }) =>
  filmSummary.Poster &&
  filmSummary.Poster !== "N/A" && (
    <Image
      source={{ uri: filmSummary.Poster }}
      style={detailStyles.poster}
      resizeMode={Image.resizeMode.contain}
    />
  )
Poster.propTypes = {
  filmSummary: PropTypes.object.isRequired,
}

const FilmPlot = ({ filmDetails }) => {
  const { plot } = filmDetails

  if (!detailExists(plot)) {
    return <Text style={detailStyles.plot}>Plot unavailable</Text>
  }
  return <Text>{filmDetails.Plot}</Text>
}
FilmPlot.propTypes = {
  filmDetails: PropTypes.object.isRequired,
}

const getDetails = filmDetails => [
  { text: `Directed by ${filmDetails.Director}` },
  { text: `Written by ${filmDetails.Writer}` },
  { text: `Cast: ${filmDetails.Actors}` },
  { text: `Language: ${filmDetails.Language}` },
  { text: `Awards: ${filmDetails.Awards}` },
  { text: `Run Time: ${filmDetails.Runtime}` },
  { text: `IMDB Rating: ${filmDetails.imdbRating}/10` },
  { text: `Box Office: ${filmDetails.BoxOffice}` },
  {
    text: "Official Website",
    url: filmDetails.Website,
    cond: detailExists(filmDetails.Website),
  },
  {
    text: "IMDB page",
    url: filmDetails.imdbID ? imdbUrl(filmDetails.imdbID) : null,
  },
]

// TODO click on poster to make full-screen
export const filmDetail = props => {
  const { filmSummary, filmDetails, isFetching } = props
  const details = getDetails(filmDetails, filmSummary)
  let i = 1

  if (isFetching) {
    return <LoadingIndicator />
  } else if (!filmDetails) {
    return null
  } else {
    return (
      <View>
        <DetailHeader {...props} />
        <Poster {...props} />
        <FilmPlot {...props} />
        <List>
          {details.map(detail => (
            <OneDetail key={`item=${i++}`} detail={detail} />
          ))}
        </List>
      </View>
    )
  }
}

filmDetail.propTypes = {
  filmSummary: PropTypes.object.isRequired,
  filmDetails: PropTypes.object,
  dispatchViewList: PropTypes.func.isRequired,
  isFetching: PropTypes.bool,
}

filmDetail.defaultProps = {
  filmDetails: {},
  isFetching: false,
}

export default connect(
  state => ({
    filmSummary: state.filmSummary,
    filmDetails: state.filmDetails,
  }),
  dispatch => ({
    dispatchViewList: () => dispatch(viewList()),
  })
)(filmDetail)
