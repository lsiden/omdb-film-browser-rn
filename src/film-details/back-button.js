import React from "react"
import { Text, TouchableOpacity } from "react-native"
import PropTypes from "prop-types"
import { connect } from "react-redux"

import { detailStyles } from "styles"
import { LEFT_TRIANGLE } from "constants"
import { viewList } from "actions"

export const backButton = ({ dispatchViewList }) => (
  <TouchableOpacity onPress={dispatchViewList}>
    <Text style={detailStyles.buttonText}>{LEFT_TRIANGLE}</Text>
  </TouchableOpacity>
)
backButton.propTypes = {
  dispatchViewList: PropTypes.func.isRequired,
}

export default connect(null, dispatch => ({
  dispatchViewList: () => dispatch(viewList),
}))(backButton)
