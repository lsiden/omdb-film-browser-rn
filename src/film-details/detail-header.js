import React from "react"
import PropTypes from "prop-types"
import { Text, View } from "react-native"
import { connect } from "react-redux"

import { detailStyles } from "styles"
import BackButton from "./back-button"

export const detailHeader = props => {
  const { filmDetails } = props
  return (
    <View style={detailStyles.header}>
      <BackButton />
      <View style={detailStyles.titleWrapper}>
        <Text style={detailStyles.title}>{filmDetails.Title}</Text>
        <Text style={detailStyles.year}>{`  (${filmDetails.Year})`}</Text>
      </View>
    </View>
  )
}

detailHeader.propTypes = {
  filmDetails: PropTypes.object.isRequired,
}

export default connect(state => ({
  filmDetails: state.filmDetails,
}))(detailHeader)
