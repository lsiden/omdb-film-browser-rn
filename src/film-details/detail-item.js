import React from "react"
import PropTypes from "prop-types"
import { Linking } from "react-native"
import { ListItem } from "react-native-elements"

import { detailStyles } from "styles"
import { detailExists } from "./detail-items"

const DetailItem = ({ detail }) => {
  const { text, cond = true, url = null } = detail

  if (!cond) {
    return null
  }
  const props = {
    style: detailStyles.item,
    title: text,
    hideChevron: true,
  }

  if (detailExists(url)) {
    props.onPress = () => {
      Linking.openURL(url)
    }
    props.hideChevron = false
  }
  return <ListItem {...props} />
}

DetailItem.propTypes = {
  detail: PropTypes.object.isRequired,
}

export default DetailItem
