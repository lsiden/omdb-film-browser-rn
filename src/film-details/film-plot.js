import React from "react"
import PropTypes from "prop-types"
import { Text } from "react-native"

import { detailExists } from "./detail-items"
import { detailStyles } from "styles"

const FilmPlot = ({ filmDetails }) => {
  const { plot } = filmDetails

  if (!detailExists(plot)) {
    return <Text style={detailStyles.plot}>Plot unavailable</Text>
  }
  return <Text>{filmDetails.Plot}</Text>
}

FilmPlot.propTypes = {
  filmDetails: PropTypes.object.isRequired,
}

export default FilmPlot
