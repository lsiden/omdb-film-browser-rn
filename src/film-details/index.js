import React from "react"
import PropTypes from "prop-types"
import { connect } from "react-redux"
import { ScrollView } from "react-native"
import { List } from "react-native-elements"

import LoadingIndicator from "components/loading-indicator"
import DetailHeader from "./detail-header"
import Poster from "./poster"
import detailItems from "./detail-items"
import FilmPlot from "./film-plot"
import DetailItem from "./detail-item"

export const filmDetail = ({ filmDetails, isFetching }) => {
  let i = 1
  const genKey = () => `item-${i++}`

  if (isFetching) {
    return <LoadingIndicator />
  } else if (!filmDetails) {
    return null
  } else {
    return (
      <ScrollView>
        <DetailHeader />
        <Poster />
        <FilmPlot />
        <List>
          {detailItems(filmDetails).map(detail => (
            <DetailItem key={genKey()} detail={detail} />
          ))}
        </List>
      </ScrollView>
    )
  }
}

filmDetail.propTypes = {
  filmDetails: PropTypes.object,
  isFetching: PropTypes.bool,
}

filmDetail.defaultProps = {
  isFetching: false,
}

export default connect(state => ({
  filmDetails: state.filmDetails || {},
  isFetching: state.isFetching,
}))(filmDetail)
