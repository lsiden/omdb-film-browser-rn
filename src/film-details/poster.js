import React from "react"
import PropTypes from "prop-types"
import { Image, TouchableOpacity } from "react-native"
import { connect } from "react-redux"

import { detailStyles } from "styles"
import { viewPoster } from "actions"
import { detailExists } from "./detail-items"

export const poster = ({ uri, dispatchViewPoster }) =>
  detailExists(uri) && (
    <TouchableOpacity onPress={() => dispatchViewPoster(uri)}>
      <Image
        source={{ uri }}
        style={detailStyles.poster}
        resizeMode={Image.resizeMode.contain}
      />
    </TouchableOpacity>
  )

poster.propTypes = {
  uri: PropTypes.string,
  dispatchViewPoster: PropTypes.func.isRequired,
}

export default connect(
  state => ({
    uri: (state.filmDetails || {}).Poster,
  }),
  dispatch => ({
    dispatchViewPoster: uri => dispatch(viewPoster(uri)),
  })
)(poster)
