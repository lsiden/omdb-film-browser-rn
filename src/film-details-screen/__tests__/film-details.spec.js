import React from "react"
import renderer from "react-native-test-utils"

import { filmDetails as FilmDetails } from "../film-details"

const details = require("__test__/__fixture__/film-details.json")

jest.mock("film-details-screen/detail-header", () => "DetailHeader")
jest.mock("film-details-screen/poster", () => "Poster")
jest.mock("film-details-screen/film-plot", () => "FilmPlot")
jest.mock("film-details-screen/detail-item", () => "DetailItem")

const defaultProps = () => {
  return {
    details,
    dispatchViewList: jest.fn()
  }
}

const createWrapper = (props = {}) =>
  renderer(<FilmDetails {...{ ...defaultProps(), ...props }} />)

it("renders", () => {
  const wrapper = createWrapper()
  expect(wrapper.toJSON()).toMatchSnapshot()
})
