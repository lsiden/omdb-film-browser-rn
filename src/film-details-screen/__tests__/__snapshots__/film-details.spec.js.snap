// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<RCTScrollView
  style={
    Object {
      "backgroundColor": "#05347f",
      "padding": 10,
    }
  }
>
  <View>
    <Poster
      onPress={[Function]}
      uri="https://images-na.ssl-images-amazon.com/images/M/MV5BYzliNzI5NzgtNDZmNy00NGU4LTg2OTQtMDEzZTRiNDU3ZTI1XkEyXkFqcGdeQXVyMDUyOTUyNQ@@._V1_SX300.jpg"
    />
    <FilmPlot />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#ffffff",
            "borderColor": "#bbb",
            "borderTopWidth": 1,
            "marginTop": 20,
          },
          undefined,
        ]
      }
    >
      <DetailItem
        detail={
          Object {
            "text": "Directed by Gillo Pontecorvo",
          }
        }
      />
      <DetailItem
        detail={
          Object {
            "text": "Written by Franco Solinas, Franco Solinas (story), Gillo Pontecorvo (story)",
          }
        }
      />
      <DetailItem
        detail={
          Object {
            "text": "Cast: Jean Martin, Yacef Saadi, Brahim Hadjadj, Tommaso Neri",
          }
        }
      />
      <DetailItem
        detail={
          Object {
            "text": "Language: French, Arabic, English",
          }
        }
      />
      <DetailItem
        detail={
          Object {
            "text": "Awards: Nominated for 3 Oscars. Another 9 wins & 5 nominations.",
          }
        }
      />
      <DetailItem
        detail={
          Object {
            "text": "Run Time: 121 min",
          }
        }
      />
      <DetailItem
        detail={
          Object {
            "text": "IMDB Rating: 8.1/10",
          }
        }
      />
      <DetailItem
        detail={
          Object {
            "text": "Box Office: $55,908",
          }
        }
      />
      <DetailItem
        detail={
          Object {
            "cond": true,
            "text": "Official Website",
            "url": "http://rialtopictures.com/battle.html",
          }
        }
      />
      <DetailItem
        detail={
          Object {
            "text": "IMDB page",
            "url": "https://www.imdb.com/title/tt0058946",
          }
        }
      />
      <DetailItem
        detail={
          Object {
            "text": "IMDB ID: tt0058946",
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;
